### ISSUE TITLE
Fix `extract_batch_size` for string values and custom batches

### ISSUE THREAD

**Issue by rohitgr7** [PyTorchLightning/pytorch-lightning#10458]
## Proposed refactoring or deprecation

String values and custom batches are not handled correctly while extracting batch sizes.

### Motivation and Pitch

**For string values:**
Currently when we try to infer the batch_size to be used for accumulating the metrics, for string values we do this:
https://github.com/PyTorchLightning/pytorch-lightning/blob/aad86423f731d1cbc175291825da161c71123006/pytorch_lightning/utilities/data.py#L33-L34
when torch dataset returns a string, the dataloader just convert it to a list of strings and since we iterate over the list and once reach the point where we have a string we just return its len, which is not correct. More discussions here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10349

Solution 1 (I prefer):
return None and raise the warning if multiple batch sizes are found.

Solution 2:
check all the values of the iterable and see if all are strings, if they are return the len of the list. This one might increase the time complexity a little bit.

---

**For custom batches:**
we return `1`
https://github.com/PyTorchLightning/pytorch-lightning/blob/aad86423f731d1cbc175291825da161c71123006/pytorch_lightning/utilities/data.py#L41-L42
which is super silent and highly inaccurate IMO. For this one too we should return None and raise the appropriate warning.

### Additional context

As always open to suggestion/discussion/thoughts and approvals :)

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

**Comment by immanuelweber** [comment]
Coming from issue https://github.com/PyTorchLightning/pytorch-lightning/issues/10349 I'd also like to add here that also lists of other types than strings would end up throwing a warning, like in my case with pathlib's Paths.
The proposed return of None and also throwing an error is also somehow inconvenient, since I can't see why a list with the length of the batch size should result in a warning, since this is what would happen since it goes down recursively until an "unkown" type comes along.
Don't get me wrong, but I feel that this consistency check can be arbitrarily complex depending on the batch structure and would always make assumptions that may not hold...
This list of value types thing could for example be solved with a check of the length of the list itself, but in some cases there might be an outer list of size batch size that contains individual lists again that are sample depended.

**Comment by rohitgr7** [comment]
@immanuelweber  we can't say for sure what the list might consists of.. it can be strings/objects/lists or anything so we can't say for sure whether it's the actual batch_size by just returning the length of the list. we need to go deeper and find a tensor which sounds like a reliable solution to me. Before the warning was added a lot of users never complaint so looks like instead of just add more complexity to it which can lead to incorrect results, raising a warning seems more reliable to me. But you can always disable this check and the warning by setting `self.log(batch_size=...)`.

**Comment by immanuelweber** [comment]
@rohitgr7 that's exactly the point, one can not handle every possible scenario, why then introduce warnings that can handle some of them but requires others to disable it, even though the scenario is perfectly valid. Why not just check the tensors and ignore everything else, if this check is actually required. My feeling is that with this new warning there is complexity added.

**Comment by rohitgr7** [comment]
yeah.. my suggestion is just to check the tensors and iterables containing tensors, raise a warning if multiple batch sizes are found and raise an error if no batch_size can be inferred. updated the description.

**Comment by immanuelweber** [comment]
mh ok. btw I don't see how I can prevent this warning by passing batch_size to the log calls. I have it in all of them and even removed all of them for comparison, but this warning is raised regardless.

**Comment by rohitgr7** [comment]
will be fixed by this: https://github.com/PyTorchLightning/pytorch-lightning/pull/10408

**Comment by Nesqulck** [comment]
You could always pass in the len of the previous list (or in a more general case the len of any container) for the next recursive call of _extract_batch_size. If you find an unknown type (e.g. strings or paths or anything other than tensors) you could just simply use the len of the previous list as the batch size. You could set this len argument as None in the constructor wich would guarantee that in cases where there are just plain unknown objects, None is retuned or a warning. I guess, this would allow to infer batch sizes from any containers such as lists, arrays, tensors, etc.

**Comment by Shreeyak** [comment]
I feel that if an Iterable is encountered, it should be ignored and not used to calculate batch size. Because of the arbitrary complexity involved, as outlined by others. I should think most of the time, if a list is being returned, the user is likely to be advanced enough to notice differing batch sizes.

I faced this warning today after updating to PL 1.5. I'm returning a list of strings of the path to ground truth meshes (len of list equal to batch size). Was not able to figure out where this warning was being raised until I came across this. It's quite annoying to see. I've run some tests and checked that the batch size of all the values I'm returning is the same.

**Comment by mpvenkatesh** [comment]
FWIW, a very large number of use cases may be captured by simply using ` yield len(sample) ` instead of `yield from _extract_batch_size(sample)`. The structure that the recursive call enforces is far more arbitrary than simply saying that if you give me a list of heterogenous constituents as being a batch, I am going to assume that you must have batch_size number of each of those constituents.

**Comment by rohitgr7** [comment]
@mpvenkatesh this is exactly what we had before, but it open up some loop-holes where user can easily trick this and we might end up not warning the user and training will continue with incorrect batch_size. We will cache this batch_size once extracted so that it won't be called again and again for each `self.log` call.

**Comment by samvanstroud** [comment]
Hi @rohitgr7, what do you suggest for those who do their own reduction on the loss before logging it? It seems confusing to suggest that the batch size is actually equal to `1` in this case in order to suppress the warning.

**Comment by rohitgr7** [comment]
@samvanstroud if you are doing your own reduction of the loss I assume you must be logging inside `_epoch_end` hooks with `on_step=False`? If that's the case, it should infer the batch_size at all, thus no warning should be there. Are you facing the warning in such a case?

**Comment by samvanstroud** [comment]
Thanks @rohitgr7, actually, I just take the mean over each batch, and then log that in `_step`.

**Comment by rohitgr7** [comment]
> Thanks @rohitgr7, actually, I just take the mean over each batch, and then log that in `_step`.

are you logging it with `on_epoch=True`?

**Comment by samvanstroud** [comment]
No, because the epochs are long enough that's its very useful to see how the loss is evolving at each step

**Comment by rohitgr7** [comment]
then it won't extract the batch_size and you shouldn't get a warning.!

**Comment by samvanstroud** [comment]
The [docs](https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#automatic-logging) suggest that batch size is used for `on_step=True`?

> `batch_size`: Current batch_size used for accumulating logs logged with `on_step=True`.
> This will be directly inferred from the loaded batch, but for some data structures you might need to explicitly provide it.

I definitely saw the warning, did this behaviour change recently?

**Comment by rohitgr7** [comment]
yeah it was changed [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/10408/files). Should be there in the recent patch release.

Nice catch.. that's a mistake.. will fix.. thanks for pointing out 😃
